### DISCUSSION TITLE
adding sessionstate provider to module got error.

### DISCUSSION THREAD

**Discussion by MRSheikh** [OrchardCMS/OrchardCore#8146] [Q&A] [answered]
hi.
i want to use AddSessionStateTempDataProvider and **not to use session with cookies** . 
i add below code in Startup.cs of my module project: 
```c#
public void ConfigureServices(IServiceCollection services)
        {
            services.AddControllers();
            services.AddMvc().AddNewtonsoftJson(options =>
                    options.SerializerSettings.ContractResolver = new DefaultContractResolver())
                  .AddSessionStateTempDataProvider();

            services.AddScoped<IResourceManifestProvider, ResourceManifest>();
            services.AddKendo();
            services.AddSession();
        }
public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }
            app.UseHttpsRedirection();
            app.UseRouting();
            app.UseAuthentication();
            app.UseAuthorization();
            app.UseEndpoints(endpoints => { endpoints.MapControllers(); });
            app.UseSession();
        }
```
**but this error occurred:**

```
An unhandled exception occurred while processing the request.
InvalidOperationException: Session has not been configured for this application or request.
Microsoft.AspNetCore.Http.DefaultHttpContext.get_Session()
```
![image](https://user-images.githubusercontent.com/17869166/103066473-3c2b9200-45ce-11eb-955b-ad33cb24bf06.png)

plz help.

**Comment by MRSheikh** [comment]
no one has idea???

**Comment by hishamco** [comment]
Seems you have issue with ordering, could try:

- Add `AddSession()` at the beginning

- Add `UseSession()` before `UseEndpoints()`

**Reply by MRSheikh** [reply]
> 
> 
> Seems you have issue with ordering, could try:
> 
>     * Add `AddSession()` at the beginning
> 
>     * Add `UseSession()` before `UseEndpoints()`

i used your tip and when i set value , error occurred:

![image](https://user-images.githubusercontent.com/17869166/103156273-50cd7d00-47bc-11eb-989f-ece38922911a.png)

**Comment by Skrypt** [comment]
I don't think you should register this from a module. It needs to be registered before `.app.UseMvc()`
Modules are probably registered after this "without looking". Try adding this to your website project startup.cs file before we register OC with `app.UseOrchardCore();`

This is a service order registration issue.

**Reply by MRSheikh** [reply]
> 
> 
> I don't think you should register this from a module. It needs to be registered before `.app.UseMvc()`
> Modules are probably registered after this "without looking". Try adding this to your website project startup.cs file before we register OC with `app.UseOrchardCore();`
> 
> This is a service order registration issue.

i add that in Startup.cs file of main site, but error exist.

**Reply by jtkech** [reply]
@MRSheikh 

> i add below code in Startup.cs of my module project:

It doesn't look like a regular module startup, look at our modules startups, they inherit from `: StartupBase` and e.g. the signature of the `Configure()` method is not the same.

    public virtual void Configure(IApplicationBuilder app, IEndpointRouteBuilder routes, IServiceProvider serviceProvider)


 Yes, at some point we allowed such raw startup for a module that doesn't inherit from `StartupBase`, but even your `Configure()` method can have a different signature, you can only use one of the above parameter, e.g. not `IWebHostEnvironment`.

So first, for testing i suggest that you use  a regular module startup as it is done in our modules

Then, in a module startup which is executed in a tenant container, most of the things are already done for you and in the right order, so they are useless here and some should not be done in a module startup as `UseRouting()` and `UseEndpoints()`.

**ConfigureServices()**

            // This is done for you
            services.AddControllers();

            // AddMvc() is done for you but you can recall it to add specific things
            services.AddMvc().AddNewtonsoftJson(options =>
                    options.SerializerSettings.ContractResolver = new DefaultContractResolver())
                  .AddSessionStateTempDataProvider();

            services.AddScoped<IResourceManifestProvider, ResourceManifest>();
            services.AddKendo();
            services.AddSession();

**Configure()**

            // Normally we call this only through the main app
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            // Just for info we have an https module
            app.UseHttpsRedirection();

            // This is done for you when building a tenant pipeline, and a module Configure() is called in this context
            app.UseRouting();

            // This is done for you
            app.UseAuthentication();
            app.UseAuthorization();

            // This is done for you when building a tenant pipeline, and a module Configure() is called in this context
            app.UseEndpoints(endpoints => { endpoints.MapControllers(); });

            app.UseSession();

**Comment by Skrypt** [comment]
Maybe the issue is that it needs to be registered on each tenants then.

**Comment by MRSheikh** [comment, marked as answer]
none of above work for me and :
```
An unhandled exception occurred while processing the request.
InvalidOperationException: Session has not been configured for this application or request.
Microsoft.AspNetCore.Http.DefaultHttpContext.get_Session()
```

**Reply by jtkech** [reply]
Just tried by adding in the startup `ConfigureServices()` of one of our modules (`OC.Contents` for the test)

            services.AddMvc()
                .AddSessionStateTempDataProvider();
            services.AddSession();

First i got the same issue, but then it was working by adding this to the module startup `Configure()`

            builder.UseSession();

**Reply by MRSheikh** [reply]
> 
> 
> Just tried by adding in the startup `ConfigureServices()` of one of our modules (`OC.Contents` for the test)
> 
> ```
>         services.AddMvc()
>             .AddSessionStateTempDataProvider();
>         services.AddSession();
> ```
> 
> First i got the same issue, but then it was working by adding this to the module startup `Configure()`
> 
> ```
>         builder.UseSession();
> ```

thanks a lot. it's work fine.
